File: ChatGPT(예진)/Java/리눅스기본명령어구현.java
import java.io.*;
import java.nio.file.*;
import java.nio.file.attribute.*;
import java.util.*;

public class 리눅스기본명령어구현 {
    
    public static void ls(String directory, boolean showHidden, boolean longFormat, boolean humanReadable) {
        File[] files = new File(directory).listFiles();
        if (files != null) {
            for (File file : files) {
                if (!showHidden && file.isHidden()) {
                    continue;
                }
                if (longFormat) {
                    String size = humanReadable ? humanReadableSize(file.length()) : Long.toString(file.length());
                    try {
                        PosixFileAttributes attrs = Files.getFileAttributeView(file.toPath(), PosixFileAttributeView.class).readAttributes();
                        System.out.printf("%s %d %d %s %s %s%n",
                                PosixFilePermissions.toString(attrs.permissions()),
                                attrs.owner().getName(), attrs.group().getName(),
                                size, attrs.lastModifiedTime(), file.getName());
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                } else {
                    System.out.println(file.getName());
                }
            }
        }
    }

    public static void cat(String filename, boolean number, Integer start, Integer end) {
        try (BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            List<String> lines = new ArrayList<>();
            String line;
            while ((line = reader.readLine()) != null) {
                lines.add(line);
            }
            if (start == null) {
                start = 0;
            }
            if (end == null) {
                end = lines.size();
            }
            for (int i = start; i < Math.min(end, lines.size()); i++) {
                if (number) {
                    System.out.printf("%d: %s%n", i + 1, lines.get(i));
                } else {
                    System.out.println(lines.get(i));
                }
            }
        } catch (IOException e) {
            System.out.printf("%s: 해당 파일 또는 디렉토리가 없습니다%n", filename);
        }
    }

    public static void echo(String text, String file, boolean append) {
        if (file != null) {
            try (BufferedWriter writer = new BufferedWriter(new FileWriter(file, append))) {
                writer.write(text);
            } catch (IOException e) {
                e.printStackTrace();
            }
        } else {
            System.out.println(text);
        }
    }

    private static String humanReadableSize(long size) {
        String[] units = {"B", "KB", "MB", "GB", "TB"};
        int index = 0;
        while (size >= 1024 && index < units.length - 1) {
            size /= 1024;
            index++;
        }
        return String.format("%.2f %s", size, units[index]);
    }

    public static void main(String[] args) {
        // ls 명령어 테스트
        System.out.println("ls 명령어 실행 결과:");
        ls(".", true, true, true);

        // cat 명령어 테스트
        System.out.println("\ncat 명령어 실행 결과:");
        cat("example.txt", true, 2, 4);

        // echo 명령어 테스트
        System.out.println("\necho 명령어 실행 결과:");
        echo("Hello, world!\nThis is a test.\n", "output.txt", true);
        echo("Another line", "output.txt", true);
        cat("output.txt", false, null, null);
    }
}

File: ChatGPT(예진)/Cpp/리눅스 기본 명령어 구현.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <sstream>
#include <sys/stat.h>
#include <windows.h>

using namespace std;

void ls(const string& directory, int show_hidden, int long_format, int human_readable);
void cat(const string& filename, int number, int start, int end);
void echo(const string& text, const string& file, int append);
void human_readable_size(double size, string& result);

int main() {
    // ls 명령어 테스트
    cout << "ls 명령어 실행 결과:" << endl;
    ls(".", 1, 1, 1);

    // cat 명령어 테스트
    cout << "\ncat 명령어 실행 결과:" << endl;
    cat("example.txt", 1, 2, 4);

    // echo 명령어 테스트
    cout << "\necho 명령어 실행 결과:" << endl;
    echo("Hello, world!\nThis is a test.\n", "output.txt", 1);
    echo("Another line", "output.txt", 1);
    cat("output.txt", 0, 0, 0);

    return 0;
}

void ls(const string& directory, int show_hidden, int long_format, int human_readable) {
    WIN32_FIND_DATAA findFileData;
    HANDLE hFind = INVALID_HANDLE_VALUE;
    char szDir[MAX_PATH];
    snprintf(szDir, sizeof(szDir), "%s\\*", directory.c_str());

    hFind = FindFirstFileA(szDir, &findFileData);

    if (hFind == INVALID_HANDLE_VALUE) {
        cout << "Failed to find files in directory." << endl;
        return;
    }

    do {
        if (!show_hidden && (findFileData.cFileName[0] == '.' || (findFileData.dwFileAttributes & FILE_ATTRIBUTE_HIDDEN))) {
            continue;
        }

        string path = directory + "\\" + findFileData.cFileName;

        struct stat file_stat;
        stat(path.c_str(), &file_stat);

        if (long_format) {
            string size_str;
            if (human_readable) {
                human_readable_size(file_stat.st_size, size_str);
            }
            else {
                stringstream ss;
                ss << file_stat.st_size;
                size_str = ss.str();
            }
            cout << oct << file_stat.st_mode << " " << file_stat.st_nlink << " " << file_stat.st_uid << " " << file_stat.st_gid << " " << size_str << " " << file_stat.st_mtime << " " << findFileData.cFileName << endl;
        }
        else {
            cout << findFileData.cFileName << endl;
        }
    } while (FindNextFileA(hFind, &findFileData) != 0);

    FindClose(hFind);
}

void cat(const string& filename, int number, int start, int end) {
    ifstream file(filename);
    if (!file.is_open()) {
        cerr << "Failed to open file: " << filename << endl;
        return;
    }

    string line;
    int line_num = 1;
    while (getline(file, line)) {
        if (line_num >= start && line_num <= end) {
            if (number) {
                cout << line_num << ": " << line << endl;
            }
            else {
                cout << line << endl;
            }
        }
        if (line_num >= end) {
            break;
        }
        line_num++;
    }

    file.close();
}

void echo(const string& text, const string& file, int append) {
    ofstream output_file(file, append ? ios::app : ios::out);
    if (!output_file.is_open()) {
        cerr << "Failed to open file: " << file << endl;
        return;
    }

    output_file << text;
    output_file.close();
}

void human_readable_size(double size, string& result) {
    const char* units[] = { "B", "KB", "MB", "GB", "TB" };
    int index = 0;

    while (size >= 1024 && index < sizeof(units) / sizeof(units[0]) - 1) {
        size /= 1024;
        index++;
    }

    stringstream ss;
    ss << fixed << size << " " << units[index];
    result = ss.str();
}
